api/datastore/mysql/mysql.go

```diff
@@ -0,0 +1,615 @@
+	res := []*models.App{}
+	filterQuery, args := buildFilterAppQuery(filter)
+	rows, err := ds.db.Query(fmt.Sprintf("SELECT DISTINCT * FROM apps %s", filterQuery), args...)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var app models.App
+		err := scanApp(rows, &app)
+
+		if err != nil {
+			if err == sql.ErrNoRows {
+				return res, nil
+			}
+			return res, err
+		}
+		res = append(res, &app)
+	}
+
+	if err := rows.Err(); err != nil {
+		return res, err
+	}
+	return res, nil
+}
+
+/*
+InsertRoute ...
+*/
+func (ds *MySQLDatastore) InsertRoute(ctx context.Context, route *models.Route) (*models.Route, error) {
+	hbyte, err := json.Marshal(route.Headers)
+	if err != nil {
+		return nil, err
+	}
+
+	cbyte, err := json.Marshal(route.Config)
+	if err != nil {
+		return nil, err
+	}
+
+	err = ds.Tx(func(tx *sql.Tx) error {
+		r := tx.QueryRow(`SELECT 1 FROM apps WHERE name=?`, route.AppName)
+		if err := r.Scan(new(int)); err != nil {
+			return models.ErrAppsNotFound
+		}
+		same, err := tx.Query(`SELECT 1 FROM routes WHERE app_name=? AND path=?`,
+			route.AppName, route.Path)
+		if err != nil {
+			return err
+		}
+		defer same.Close()
+		if same.Next() {
+			return models.ErrRoutesAlreadyExists
+		}
+
+		_, err = tx.Exec(`
+		INSERT INTO routes (
+			app_name,
+			path,
+			image,
+			format,
+			maxc,
+			memory,
+			type,
+			timeout,
+			headers,
+			config
+		)
+		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?);`,
+			route.AppName,
+			route.Path,
+			route.Image,
+			route.Format,
+			route.MaxConcurrency,
+			route.Memory,
+			route.Type,
+			route.Timeout,
+			string(hbyte),
+			string(cbyte),
+		)
+		return err
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	return route, nil
+}
+
+/*
+UpdateRoute ...
+*/
+func (ds *MySQLDatastore) UpdateRoute(ctx context.Context, newroute *models.Route) (*models.Route, error) {
+	var route models.Route
+	err := ds.Tx(func(tx *sql.Tx) error {
+		row := ds.db.QueryRow(fmt.Sprintf("%s WHERE app_name=? AND path=?", routeSelector), newroute.AppName, newroute.Path)
+		if err := scanRoute(row, &route); err == sql.ErrNoRows {
+			return models.ErrRoutesNotFound
+		} else if err != nil {
+			return err
+		}
+
+		route.Update(newroute)
+
+		hbyte, err := json.Marshal(route.Headers)
+		if err != nil {
+			return err
+		}
+
+		cbyte, err := json.Marshal(route.Config)
+		if err != nil {
+			return err
+		}
+
+		res, err := tx.Exec(`
+		UPDATE routes SET
+			image = ?,
+			format = ?,
+			maxc = ?,
+			memory = ?,
+			type = ?,
+			timeout = ?,
+			headers = ?,
+			config = ?
+		WHERE app_name = ? AND path = ?;`,
+			route.Image,
+			route.Format,
+			route.MaxConcurrency,
+			route.Memory,
+			route.Type,
+			route.Timeout,
+			string(hbyte),
+			string(cbyte),
+			route.AppName,
+			route.Path,
+		)
+
+		if err != nil {
+			return err
+		}
+
+		if n, err := res.RowsAffected(); err != nil {
+			return err
+		} else if n == 0 {
+			return models.ErrRoutesNotFound
```

Comment:
`ErrRoutesNotFound` should be caught above, so affecting 0 rows might be worth logging here since it would indicate something pretty severely wrong like a bug or DB integrity problems.